api/envoy/api/v2/listener/listener.proto

```diff
@@ -87,6 +87,16 @@ message FilterChainMatch {
   // listener in determining a filter chain match.
   // [#not-implemented-hide:]
   google.protobuf.UInt32Value destination_port = 8;
+
+  // If non-empty, a transport protocol to consider when determining a filter chain match.
+  // This value will be compared against transport protocol of a new connection, when it's
+  // detected by one of the listener filters (this needs to be configured separately).
+  //
+  // Valid values include:
+  // * "raw_buffer" - default, used when no transport protocol is detected,
+  // * "tls" - set by :ref:`envoy.listener.tls_inspector <config_listener_filters_tls_inspector>`
+  //   when TLS protocol is detected.
+  string transport_protocol = 9;
```

Comment:
I think in general, you can imagine listener filters setting arbitrary key/value and wanting to match here. In that sense, `{transport_protocol=tls, sni_domains=[foo.com]}` is one example. Does it make sense to generalize to a `Struct` match?

Reply:
This is a bit more complex than a simple key-value matching, i.e. options covered in this PR, and the one I'm going to do next (destination IP:port), are matching on:
- transport protocol,
- server names / sni domains,
- destination IP:port.

Those are not equal, and we need to filter on destination IP:port first, then on server names (or use catch-all), then on transport protocol (or use catch-all).

...and even that isn't totally obvious, i.e. should `{requestedServerName=www.example.com, detectedTransportProtocol=tls}` be matched against `{sni_domains=[www.example.com],transport_protocol=}` or `{sni_domains=,transport_protocol=tls}`?

I'm in favor of the former, since it better matches "virtual hosts" concept, but I could see the arguments being made for the either of those.

Reply:
Yeah, I agree that the full logic can't be captured as a simple key/value match. I'm curious though if you think it makes sense to allow listener filters more expressive power than just a string match on protocol here? We could punt this to future work with a TODO if it makes sense.